framework/frequency_manager.hpp

```diff
@@ -0,0 +1,220 @@
+    int current_set_frequency;
+    std::vector<int> frequency_levels;
+    int frequency_level_idx = 0;
+    static constexpr int total_frequency_levels = 9;
+#endif
+
+    std::string read_file(const std::string &file_path)
+    {
+#ifdef __linux__
+        /* Read first line of given file */
+        char line[100]; //100 characters should be more than enough
+        FILE *file = fopen(file_path.c_str(), "r");
+
+        if (file == NULL) {
+            fprintf(stderr, "%s: cannot read from file: %s :%m\n", program_invocation_name, file_path.c_str());
+            exit(EXIT_FAILURE);
+        }
+        fscanf(file, "%s", line);
+        fclose(file);
+        return std::string(line);
+#else
+        return "";
+#endif
+    }
+
+    int write_file(const std::string &file_path, const std::string &line)
+    {
+#ifdef __linux__
+        FILE *file = fopen(file_path.c_str(), "w");
+
+        if (file == NULL)
+            return 1;
+
+        fprintf(file, "%s", line.c_str());
+        fclose(file);
+#endif
+        return 0;
+    }
+
+    int get_frequency_from_file(const std::string &file_path)
+    {
+        /* Read frequency value from file */
+        int frequency;
+#ifdef __linux__
+        FILE *file = fopen(file_path.c_str(), "r");
+
+        if (file == NULL) {
+            fprintf(stderr, "%s: cannot read from file: %s :%m\n", program_invocation_name, file_path.c_str());
+            exit(EXIT_FAILURE);
+        }
+        fscanf(file, "%d", &frequency);
+        fclose(file);
+#endif
+        return frequency;
+    }
+
+    void populate_frequency_levels()
+    {
+#ifdef __linux__
+        frequency_levels.push_back(max_frequency_supported);
+        frequency_levels.push_back(min_frequency_supported); 
+
+        std::vector<int> tmp = frequency_levels;
+
+        while (frequency_levels.size() != total_frequency_levels)
+        {
+            std::sort(tmp.begin(), tmp.end(), std::greater<int>());
+            for (int idx = 1; idx < tmp.size(); idx++)
+                frequency_levels.push_back((tmp[idx] + tmp[idx - 1]) / 2);
+            tmp = frequency_levels;
+        }
+#endif
+    }
+
+    void check_if_userspace_present()
+    {
+#ifdef __linux__
+        const char *scaling_governor_path = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_available_governors";
+        char read_file[100];
+        FILE *file = fopen(scaling_governor_path, "r");
+
+        if (file == NULL) {
+            fprintf(stderr, "%s: cannot read from file: %s :%m\n", program_invocation_name, scaling_governor_path);
+            exit(EXIT_FAILURE);
+        }
+
+        while (fscanf(file, "%s", read_file)) {
+            if (strcmp(read_file, "userspace") == 0) {
+                fclose(file);
+                return;
+            }
+        }
+
+        fprintf(stderr, "%s: Cannot find \"userspace\" scaling governor from the file: %s\n", program_invocation_name, scaling_governor_path);
+        exit(EXIT_FAILURE);
```

Comment:
`EXIT_FAILURE` is not acceptable in the FW because we use it exclusively to mean we've found an SDE. Use something else. Repeats.

Reply:
Changed 3 of them to `EXIT_IOERR` since the issue is not not able to open/write to file. And the other one `EX_DATAERR`.